File: haskell solutions/problem_17.hs
--Program that adds up the letters from numbers in the range 1-1000
--Spaces and hyphens don't cound and the numbers are written in compliance with the British usage
--ex: 342 - three hundred and forty-two has 23 letters

one_digit = [0, 3, 3, 5, 4, 4, 3, 5, 5, 4]      -- from 0 to 9
teens = [3, 6, 6, 8, 8, 7, 7, 9, 8, 8]          -- from 10 to 19
tens = [0, 0, 6, 6, 5, 5, 5, 7, 6, 6]           -- 20, 30, ..., 90

_and = 3
thousand = 8
hundred = 7

calculate_hundreds::Int -> Int
calculate_hundreds number = 
    if (number `rem` 100) `div` 10 == 0 && (number `rem` 100) `rem` 10 == 0 --ex: 100, 200, 300
        then one_digit !! (number `div` 100) + hundred
    else if (number `rem` 100) `div` 10 == 1                                --ex: 110, 111, ..., 119
        then one_digit !! (number `div` 100) + hundred + _and + teens !! ((number `rem` 100) `rem` 10)
    else one_digit !! (number `div` 100) + hundred + _and + tens !! ((number `rem` 100) `div` 10) + one_digit !! ((number `rem` 100) `rem` 10)                                            --this is for everything else


compute::Int -> Int -> Int
compute n m = 
    let result_list = map (\x -> if x < 10 then one_digit !! (x `rem` 10)
                    else if x == 10 then 3
                    else if x < 20 then teens !! (x `rem` 10)
                    else if x < 100 then tens !! (x `div` 10) +  one_digit !! (x `rem` 10)
                    else calculate_hundreds x) [n..m]
    in sum result_list

main::IO()
main = do 
    let result = (compute 1 999) + 11       --plus one thousand 
    print result

File: python solutions/problem_17.py
# Program that adds up the letters from numbers in the range 1-1000
# Spaces and hyphens don't cound and the numbers are written in compliance with the British usage
# ex: 342 - three hundred and forty-two has 23 letters

# zero, one, two, three, four, five, six, seven, eight, nine
one_digit = {0: 0, 1: 3, 2: 3, 3: 5, 4: 4, 5: 4, 6: 3, 7: 5, 8: 5, 9: 4}

# eleven, twelve, thirteen, fourteen, fifteen, sixteen, seventeen, eighteen, nineteen
teens = {10: 3, 11: 6, 12: 6, 13: 8, 14: 8, 15: 7, 16: 7, 17: 9, 18: 8, 19: 8}

# twenty, thirty, forty, fifty, sixty, seventy, eighty, ninety
tens = {0: 0, 20: 6, 30: 6, 40: 5, 50: 5, 60: 5, 70: 7, 80: 6, 90: 6}

_hundred = 7
_and = 3


def add_up_letters(i, n):
    sum_of_letters = 0

    for number in range(i, n + 1):
        if len(str(number)) == 1:
            sum_of_letters += one_digit[number]
        if len(str(number)) == 2:
            if number in teens:
                sum_of_letters += teens[number]
            else:
                sum_of_letters += one_digit[number % 10]
                sum_of_letters += tens[number - number % 10]

        if len(str(number)) == 3:
            if number == 100:
                sum_of_letters += 10
            else:
                sum_of_letters += one_digit[number // 100] + _hundred
            if number % 100 in teens:
                sum_of_letters += teens[number % 100]
                sum_of_letters += _and
            else:
                sum_of_letters += one_digit[number % 10]
                if number - number % 10 - number // 100 * 100 == 0 and number % 10 == 0:
                    pass
                else:
                    sum_of_letters += _and
                    sum_of_letters += tens[number - number % 10 - number // 100 * 100]

        if len(str(number)) == 4:
            sum_of_letters += len("onethousand")

    return sum_of_letters


if __name__ == "__main__":
    print(add_up_letters(1, 1000))
